### DISCUSSION TITLE
Cross testing in wine (linux host, windows target) fails with nightly rust

### DISCUSSION THREAD

**Discussion by VorpalBlade** [cross-rs/cross#1386] [Q&A] [answered]
As of rustc 1.76.0-nightly (a96d57bdb 2023-12-15) cross compiling from a Linux host to Windows fails to run some of my tests due to Wine not implementing functionality. This works fine on stable toolchain (1.74).

Is there a way to tell cross to try using a newer wine or something to see if this is fixed?

The errors look like this:

```
---- add::tests::check_add_normal_script stdout ----
thread 'add::tests::check_add_normal_script' panicked at src/add/tests.rs:301:6:
called `Result::unwrap()` on an `Err` value: Call not implemented. (os error 120)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- add::tests::check_add_smart_script stdout ----
thread 'add::tests::check_add_smart_script' panicked at src/add/tests.rs:325:6:
called `Result::unwrap()` on an `Err` value: Call not implemented. (os error 120)

```

Unfortunately setting RUST_BACKTRACE=1 doesn't seem to propagate through cross into whatever container it runs the test in.

Here is a CI run showing what happens: https://github.com/VorpalBlade/chezmoi_modify_manager/actions/runs/7231476998/job/19704690937

The same happens locally (where I use podman instead of docker) with: `cross +nightly test --config 'build.rustc-wrapper=""' --config 'target.x86_64-unknown-linux-gnu.linker="/usr/bin/gcc"'  --target x86_64-pc-windows-gnu`

(extra flags are to turn off sccache and mold which cross doesn't seem to like)

**Comment by Emilgardis** [comment]
You can figure out what change broke this with [cargo-bisect-rustc](https://github.com/rust-lang/cargo-bisect-rustc) using a script like 


```bash
#!/usr/bin/env bash
export CROSS_CUSTOM_TOOLCHAIN=1
cross test --target x86_64-pc-windows-gnu
```

and then run

```
cargo bisect-rustc --script bisect.sh --target x86_64-pc-windows-gnu --host x86_64-unknown-linux-gnu --end 2023-12-16 --start 2023-10-02
```

I recently upgraded wine on the main branch images, https://github.com/cross-rs/cross/pull/1374, altough when I test it with the main images it still fails with the same error 

I'll bisect and return with the result

**Reply by Emilgardis** [reply, marked as answer]
searched nightlies: from nightly-2023-12-13 to nightly-2023-12-15
regressed nightly: nightly-2023-12-14
searched commit range: https://github.com/rust-lang/rust/compare/3340d49d22b1aba425779767278c40781826c2f5...eeff92ad32c2627876112ccfe812e19d38494087
regressed commit: https://github.com/rust-lang/rust/commit/77d16997566dfb2384bcbc504e2579c7ae973666

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --script bisect.sh --target x86_64-pc-windows-gnu --host x86_64-unknown-linux-gnu --end 2023-12-15 --start 2023-12-13 
```

```bash
#!/usr/bin/env bash
export CROSS_CUSTOM_TOOLCHAIN=1
export CUSTOM_TOOLCHAIN_COMPAT=nightly-x86_64-unknown-linux-gnu
cross test --target x86_64-pc-windows-gnu -v &>> log.txt
```
</details>

**Reply by Emilgardis** [reply]
I've made a comment on https://github.com/rust-lang/rust/pull/116438#issuecomment-1859000492, I suspect a real report should be made though

**Reply by Emilgardis** [reply]
In https://github.com/rust-lang/rust/pull/116438#issuecomment-1859224719 it's pointed out that the issue is with `FileAllocationInformation` not being handled here https://gitlab.winehq.org/wine/wine/-/blob/master/dlls/ntdll/unix/file.c?ref_type=heads#L4566

I think a report should be done for wine on their bugtracker

**Reply by workingjubilee** [reply]
Please do file this with Wine.

**Reply by VorpalBlade** [reply]
Thanks, I will do so when I get some time this weekend.

Thanks a lot for going above and beyond with debugging this. I was just hoping for some general pointers as to where to even begin (since I'm largely unfamiliar with modern Windows).

**Reply by Emilgardis** [reply]
No problem! A little pointer also, https://github.com/rust-lang/rust/pull/119051 was merged which should fix the immediate issue with `write(true).create(true).truncate(true)`, but the actual missing implementation is still an issue, just not as critical for rust as it's not directly accessible anymore.
